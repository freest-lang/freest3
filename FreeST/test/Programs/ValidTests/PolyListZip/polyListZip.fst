data List a = Cons a (List a) | Nil

main1 : List Int
main1 = Cons[Int] 1 $ Cons [Int] 2 $ Cons [Int] 3 $ Cons [Int] 4 $ Nil[Int]

main2 : List Char
main2 = Cons[Char] 'a' $ Cons [Char] 'b' $ Cons [Char] 'c' $ Nil[Char]

zip : ∀ a b . List a → List b → List (a,b)
zip xs ys =
  case xs of {
    Cons x xs ->
      case ys of {
        Cons y ys -> Cons[(a,b)] (x,y) (zip[a,b] xs ys),
        Nil -> Nil[(a,b)]
      },
    Nil -> Nil[(a,b)]
  }


main : List (Int, Char)
main = zip[Int, Char] main1 main2