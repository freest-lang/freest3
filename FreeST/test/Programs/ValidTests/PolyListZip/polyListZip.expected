Cons (1, 'a') (Cons (2, 'b') (Cons (3, 'c') Nil))